src/main/scala/org/broadinstitute/hail/driver/FilterAlleles.scala

```diff
@@ -0,0 +1,157 @@
+
+object FilterAlleles extends Command {
+  class Options extends BaseOptions {
+    @Args4jOption(required = false, name = "--keep", usage = "Keep variants matching condition")
+    var keep: Boolean = false
+
+    @Args4jOption(required = false, name = "--remove", usage = "Remove variants matching condition")
+    var remove: Boolean = false
+
+    @Args4jOption(required = true, name = "-c", aliases = Array("--condition"),
+      usage = "Filter expression involving v (variant), va (variant annotations), and aIndex (allele index)",
+      metaVar = "COND_EXPR")
+    var condition: String = _
+
+    @Args4jOption(required = true, name = "-a", aliases = Array("--annotation"),
+      usage = "Annotation modifying expression involving v (new variant), va (old variant annotations), and aIndices (maps from new to old indices)",
+      metaVar = "ANNO_EXPR")
+    var annotation: String = _
+  }
+
+  override def newOptions: Options = new Options
+
+  override def name: String = "filteralleles"
+
+  override def description: String = "Filter alleles in current dataset using the Hail expression language"
+
+  override def supportsMultiallelic: Boolean = true
+
+  override def requiresVDS: Boolean = true
+
+  override protected def run(state: State, options: Options): State = {
+    if (state.vds.wasSplit)
+      warn("this VDS was already split; this module was designed to handle multi-allelics, perhaps you should use filtervariants instead.")
+
+    if (!(options.keep ^ options.remove))
+      fatal("either `--keep' or `--remove' required, but not both")
+
+    val conditionEC = EvalContext(Map(
+      "v" -> (0, TVariant),
+      "va" -> (1, state.vds.vaSignature),
+      "aIndex" -> (2, TInt)))
+    val conditionE = Parser.parse[Boolean](options.condition, conditionEC, TBoolean)
+
+    val annotationEC = EvalContext(Map(
+      "v" -> (0, TVariant),
+      "va" -> (1, state.vds.vaSignature),
+      "aIndices" -> (2, TArray(TInt))))
+    val (types, generators) = Parser.parseAnnotationArgs(options.annotation, annotationEC, Annotation.VARIANT_HEAD)
+    val inserterBuilder = mutable.ArrayBuilder.make[Inserter]
+    val finalType = types.foldLeft(state.vds.vaSignature) { case (vas, (path, signature)) =>
+      val (newVas, i) = vas.insert(signature, path)
+      inserterBuilder += i
+      newVas
+    }
+    val inserters = inserterBuilder.result()
+
+    val keep = options.keep
+
+    def filterAllelesInVariant(v: Variant, va: Annotation): Option[(Variant, IndexedSeq[Int], Array[Int])] = {
+      var alive = 0
+      val oldToNew = new Array[Int](v.nAlleles)
+      for (aai <- v.altAlleles.indices) {
+        val index = aai + 1
+        conditionEC.setAll(v, va, index)
+        oldToNew(index) =
+          if (Filter.keepThis(conditionE(), keep)) {
+            alive += 1
+            alive
+          } else
+            0
+      }
+
+      if (alive == 0)
+        None
+      else {
+        val survivorIndexMap = oldToNew.iterator
```

Comment:
You call this a bunch of different things in the code: `survivorIndexMap`, `survivingIndexMap`, `indexMap` and `aIndices`.  Pick one.  I'm partial to `newToOld`.